src/editor/editorinteractive.cc

```diff
@@ -951,6 +1287,621 @@ void EditorInteractive::map_changed(const MapWas& action) {
+			} else if ((*c >= 'a' && *c <= 'z') || (*c >= '0' && *c <= '9')) {
+				textdomain += *c;
+			} else if (*c >= 'A' && *c <= 'Z') {
+				textdomain += (*c + 'a' - 'A');
+			}
+		}
+		if (textdomain.empty()) {
+			log("WARNING: Map name '%s' unsuited for creating a set_textdomain() argument\n", mapname);
+			textdomain = "invalid_name";
+		}
+		fw.print_f("set_textdomain(\"scenario_%s.wmf\")\n", textdomain.c_str());
+	}
+
+	// Builtin includes
+	{
+		// We always include this one because it defines run()
+		std::set<std::string> builtin_includes = {"scripting/coroutine.lua"};
+		// Check which other includes to include
+		for (const FS_FunctionCall* f : scripting_saver_->all<FS_FunctionCall>()) {
+			if (const BuiltinFunctionInfo* b = builtin_f(*f->get_function())) {
+				if (!b->included_from.empty() && !builtin_includes.count(b->included_from)) {
+					builtin_includes.insert(b->included_from);
+				}
+			}
+		}
+		fw.string("\n");
+		for (const std::string& i : builtin_includes) {
+			fw.print_f("include \"%s\"\n", i.c_str());
+		}
+	}
+
+	// Global variables
+	if (!variables_.empty()) {
+		fw.string("\n");
+		for (const auto& var : variables_) {
+			var->write_lua(0, fw);
+		}
+	}
+
+	// User-defined functions
+	for (const auto& f : scripting_saver_->all<LuaFunction>()) {
+		f->write_lua(0, fw);
+		fw.string("\n");
+	}
+
+	// Hand-written includes
+	// NOTE: Those should not contain "directly scripted" code but only functions which
+	// can then be invoked from the user-defined functions here (not yet implemented)
+	if (!includes_.empty()) {
+		fw.string("\n");
+		for (const std::string& i : includes_) {
+			fw.print_f("include \"map:%s\"\n", i.c_str());
+		}
+	}
+
+	// Autogenerate a function to initialize map objects and stuff as the editor user laboriously
+	// defined them
+	{
+		auto generate_unused_name = [this](std::string basename) {
+			while (scripting_saver_->get<Variable>(basename) ||
+			       scripting_saver_->get<LuaFunction>(basename)) {
+				basename = "_" + basename;
+			}
+			return basename;
+		};
+
+		const std::string autogen_starting_function = generate_unused_name("autogenerated");
+		fw.print_f("\nfunction %s()\n", autogen_starting_function.c_str());
+
+#define var autogen_variable_name.c_str()
+#define write(text, ...) fw.print_f("   " text "\n", __VA_ARGS__)
+
+		const std::string autogen_variable_name = generate_unused_name("temp_object");
+		write("local %s", var);
+		const std::string autogen_map_name = generate_unused_name("map");
+		write("local %s = wl.Game().map", autogen_map_name.c_str());
+
+		const uint8_t nrplayers = map.get_nrplayers();
+		std::unique_ptr<std::string[]> pvar(new std::string[nrplayers]);
+		for (unsigned i = 1; i <= nrplayers; ++i) {
+			pvar[i - 1] = generate_unused_name("p" + std::to_string(i));
+			write("local %s = wl.Game().players[%u]", pvar[i - 1].c_str(), i);
+		}
+
+		// Map objects
+		auto stockpolicystring = [](Widelands::StockPolicy p) {
+			switch (p) {
```

Comment:
Shift `wh_policy_to_string` and `string_to_wh_policy` from `scripting/lua_map.cc` to `logic/map_objects/tribes/building_settings.h` and you can get rid of this code duplication.